pkg/nodes/list.go

```diff
@@ -60,3 +68,98 @@ func ListExternalIPv4Networks(apiClient *clients.Settings, options v1.ListOption
 
 	return ipV4ExternalAddresses, nil
 }
+
+// WaitForAllNodesAreReady waits for all nodes to be Ready for a time duration up to the timeout.
+func WaitForAllNodesAreReady(apiClient *clients.Settings,
+	timeout time.Duration,
+	options v1.ListOptions) (bool, error) {
+	glog.V(100).Infof("Waiting for all nodes in the Ready state up to duration of %v",
+		timeout)
+
+	nodesList, err := List(apiClient, options)
+	if err != nil {
+		glog.V(100).Infof("Failed to list all nodes due to %s", err.Error())
+
+		return false, err
+	}
+
+	// Wait 5 secs in each iteration before condition function () returns true or errors or times out
+	// after availableDuration
+	err = wait.PollImmediate(backoff, timeout, func() (done bool, err error) {
+		// iterate through the ClusterOperators in the list.
+		for _, node := range nodesList {
+			_, err = node.IsReady()
+			if err != nil {
+				glog.V(100).Infof("Node %v not Ready", node.Object.Name)
+
+				return false, err
+			}
+		}
+
+		// Here we are always returning "false, nil" so we keep iterating throughout the stableInterval
+		// of the inner wait.pollImmediate loop, until we time out.
+		return true, nil
+	})
+
+	if err == nil {
+		glog.V(100).Infof("All nodes were found in the Ready State during availableDuration: %v",
+			timeout)
+
+		return true, nil
+	}
+
+	// Here err is "timed out waiting for the condition"
+	glog.V(100).Infof("Not all nodes were found in the Ready State during availableDuration: %v",
+		err)
+
+	return false, err
+}
+
+// WaitForAllNodesToReboot waits for all nodes to start and finish reboot up to the timeout.
+func WaitForAllNodesToReboot(apiClient *clients.Settings,
+	options v1.ListOptions,
+	timeout time.Duration) (bool, error) {
+	glog.V(100).Infof("Waiting for all nodes the list to reboot and return to the Ready condition")
+
+	nodesList, err := List(apiClient, options)
+	if err != nil {
+		glog.V(100).Infof("Failed to list all nodes due to %s", err.Error())
+
+		return false, err
+	}
+
+	readyNodes := []string{}
+	err = wait.PollImmediate(backoff, timeout, func() (done bool, err error) {
```

Comment:
Does this work as expected? Just wondering how safe this function is as it seems like a race condition could be encountered easily

Reply:
it worked for me, all code I push was tested, at least before we start to modify it during review. But I could re-test the last code version to be sure

Reply:
if you have any idea how we could modify it to be more stable I will be more then happy to here your suggestions.

Reply:
@trewest please, take look on the last code version, I did some changes, looks much better and works just fine